---
title: "ClusteringWithSeurat"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
library(Seurat)
library(cowplot)
# read in the expression matrix
Combined = read.csv('Combined.csv',row.names=1)
dim(Combined)
# This has genes as rows (~27k) and cells as columns (1417)
# read in the annotation file
AnnotationFile = read.csv('CellAnnotation.csv',row.names=1)
# Sanity check
dim(AnnotationFile)
head(Annotation)
tail(Annotation)

```

# http://satijalab.org/seurat/interaction_vignette.html

# Now Create a seurat object
```{r}
# First make the expression data as a sparse data frame to speed up computation. 
# Our combined data is not numeric, so convert the data frame to numeric
Combined1 = data.matrix(Combined)
# Then convert to a sparse matrix
expression.data <- Matrix(Combined1, sparse = T)
# Sanity check
expression.data[1:3,1:3]

# Format the AnnotationFile
# Add a new column called CellNames and name them with sequential numbers
AnnotationFile1 = cbind(AnnotationFile,'CellNames'=1:nrow(AnnotationFile))
# Paste the string Cell to each of the numbers
AnnotationFile1$CellNames <- paste("Cell",AnnotationFile1$CellNames, sep="_")
# Now make this CellNames column as the row
rownames(AnnotationFile1) = AnnotationFile1$CellNames
# Now drop the CellNames column
library(dplyr)
AnnotationFile2 = AnnotationFile1 %>% select(-CellNames)

# Now create a seurat obj
expressionObject <- CreateSeuratObject(raw.data = expression.data, min.cells = 3)

# Add the metadata to the object
expressionObject_1 <- AddMetaData(expressionObject, metadata = AnnotationFile2)
# Filter out cells 
expressionFilteredObj <- FilterCells(expressionObject_1, subset.names = "nGene", low.thresholds = 500,  high.thresholds = Inf)
# Normalize data set 
expressionNormalized <- NormalizeData(expressionFilteredObj)
# Keep only variable genes
expressionVariable <- FindVariableGenes(expressionNormalized, x.low.cutoff = 0.1)
# Scale the data
expressionScaled <- ScaleData(expressionVariable, genes.use = expressionVariable@var.genes, model.use = "negbinom")
# Now reduce the dimensions by running the PCA and saving 30 dimensions
# So all the features (~27k genes) are collapsed into 30 features
expressionPCA <- RunPCA(expressionScaled, pcs.compute = 30, weight.by.var = FALSE)
# Use the dimensions generated by PCA to create the tsne but we dont use all the dims
# we just use first 19 dimensions
expressionTsne <- RunTSNE(expressionPCA, dims.use = 1:19, do.fast = T)
# So now lets find all the cluster markers
expressionClusters <- FindClusters(expressionTsne, reduction.type = "pca", dims.use = 1:19, save.SNN = T)
# color by cluster ID, annotated cluster from the manuscript, or batch
# Can switch the identity class using SetAllIdent if desired
TSNEPlot(expressionClusters, do.label = T)

# TSNEPlot(expressionClusters,group.by='CellType' )


```

# Use an alternative seurat work flow
```{r}
New <- CreateSeuratObject(raw.data = expression.data, min.cells = 3, min.genes = 200, 
    project = "10X_old")

mito.genes <- grep(pattern = "^MT-", x = rownames(x = New@data), value = TRUE)
percent.mito <- Matrix::colSums(New@raw.data[mito.genes, ])/Matrix::colSums(New@raw.data)

New <- AddMetaData(object = New, metadata = percent.mito, col.name = "percent.mito")
# This threw up a warning. Understand that warning. The plot is also kinda different from the plot from the tutorial;
# The warning is that all cells have the same value of feature
VlnPlot(object = New, features.plot = c("nGene", "nUMI", "percent.mito"), nCol = 3)

# We filter out cells that have unique gene counts over 2,500 or less than
# 200 Note that low.thresholds and high.thresholds are used to define a
# 'gate' -Inf and Inf should be used if you don't want a lower or upper
# threshold.
New <- FilterCells(object = New, subset.names = c("nGene", "percent.mito"), 
    low.thresholds = c(200, -Inf), high.thresholds = c(2500, 0.05))

# Normalizing the data
# After removing unwanted cells from the dataset, the next step is to normalize the data. By default, we employ a global-scaling normalization method “LogNormalize” that normalizes the gene expression measurements for each cell by the total expression, multiplies this by a scale factor (10,000 by default), and log-transforms the result.

New <- NormalizeData(object = New, normalization.method = "LogNormalize", 
    scale.factor = 10000)

# Detection of variable genes across the single cells
# Seurat calculates highly variable genes and focuses on these for downstream analysis. FindVariableGenes calculates the average expression and dispersion for each gene, places these genes into bins, and then calculates a z-score for dispersion within each bin. This helps control for the relationship between variability and average expression. This function is unchanged from (Macosko et al.), but new methods for variable gene expression identification are coming soon. We suggest that users set these parameters to mark visual outliers on the dispersion plot, but the exact parameter settings may vary based on the data type, heterogeneity in the sample, and normalization strategy. The parameters here identify ~2,000 variable genes, and represent typical parameter settings for UMI data that is normalized to a total of 1e4 molecules.

New <- FindVariableGenes(object = New, mean.function = ExpMean, dispersion.function = LogVMR, 
    x.low.cutoff = 0.0125, x.high.cutoff = 3, y.cutoff = 0.5)

# Found 3106 variable genes
length(x = New@var.genes)

#Scaling the data and removing unwanted sources of variation
#Your single cell dataset likely contains ‘uninteresting’ sources of variation. This could include not only technical noise, but batch effects, or even biological sources of variation (cell cycle stage). As suggested in Buettner et al, NBT, 2015, regressing these signals out of the analysis can improve downstream dimensionality reduction and clustering. To mitigate the effect of these signals, Seurat constructs linear models to predict gene expression based on user-defined variables. The scaled z-scored residuals of these models are stored in the scale.data slot, and are used for dimensionality reduction and clustering.

#We can regress out cell-cell variation in gene expression driven by batch (if applicable), cell alignment rate (as provided by Drop-seq tools for Drop-seq data), the number of detected molecules, and mitochondrial gene expression. For cycling cells, we can also learn a ‘cell-cycle’ score (see example [HERE]) and regress this out as well. In this simple example here for post-mitotic blood cells, we regress on the number of detected molecules per cell as well as the percentage mitochondrial gene content.

# Seurat v2.0 implements this regression as part of the data scaling process. Therefore, the RegressOut function has been deprecated, and replaced with the vars.to.regress argument in ScaleData.

New <- ScaleData(object = New, vars.to.regress = c("nUMI", "percent.mito"))

New <- RunPCA(object = New, pc.genes = New@var.genes, do.print = TRUE, pcs.print = 1:5, 
    genes.print = 5)

 #Examine and visualize PCA results a few different ways
PrintPCA(object = New, pcs.print = 1:5, genes.print = 5, use.full = FALSE)

VizPCA(object = New, pcs.use = 1:2)

PCAPlot(object = New, dim.1 = 1, dim.2 = 2)
PCHeatmap(object = New, pc.use = 1:12, cells.use = 500, do.balanced = TRUE, label.columns = FALSE, use.full = FALSE)
PCElbowPlot(object = New)

New <- FindClusters(object = New, reduction.type = "pca", dims.use = 1:10, 
    resolution = 0.6, print.output = 0, save.SNN = TRUE)

New <- RunTSNE(object = New, dims.use = 1:10, do.fast = TRUE)

TSNEPlot(object = New,do.label=TRUE)

# Add two new columns to the metadata which will group cells by age_cluster or cluster_age
New@meta.data$Cluster_Age = paste(New@meta.data$CellType,New@meta.data$res.0.6,sep='_')
New@meta.data$Age_Cluster = paste(New@meta.data$res.0.6,New@meta.data$CellType,sep='_')

# find all markers of cluster 0
cluster0.markers <- FindMarkers(object = New, ident.1 = 0, min.pct = 0.25)
print(x = head(x = cluster0.markers, n = 10))
# find all markers of cluster 1
cluster1.markers <- FindMarkers(object = New, ident.1 = 1, min.pct = 0.25)
print(x = head(x = cluster1.markers, n = 10))
# find all markers of cluster 2
cluster2.markers <- FindMarkers(object = New, ident.1 = 2, min.pct = 0.25)
print(x = head(x = cluster2.markers, n = 10))
# find all markers of cluster 3
cluster3.markers <- FindMarkers(object = New, ident.1 = 3, min.pct = 0.25)
print(x = head(x = cluster3.markers, n = 10))
cluster4.markers <- FindMarkers(object = New, ident.1 = 4, min.pct = 0.25)
print(x = head(x = cluster4.markers, n = 10))
cluster5.markers <- FindMarkers(object = New, ident.1 = 5, min.pct = 0.25)
print(x = head(x = cluster5.markers, n = 10))
cluster6.markers <- FindMarkers(object = New, ident.1 = 6, min.pct = 0.25)
print(x = head(x = cluster6.markers, n = 10))
cluster7.markers <- FindMarkers(object = New, ident.1 = 7, min.pct = 0.25)
print(x = head(x = cluster7.markers, n = 10))

# find markers for every cluster compared to all remaining cells, report
# only the positive ones
New.markers <- FindAllMarkers(object = New, only.pos = TRUE, min.pct = 0.25, 
    thresh.use = 0.25)
New.markers %>% group_by(cluster) %>% top_n(10, avg_logFC)

# find markers for every cluster compared to all remaining cells, report
# only the positive ones
New.markers <- FindAllMarkers(object = New, only.pos = TRUE, min.pct = 0.25, 
    thresh.use = 0.25)
New.markers %>% group_by(cluster) %>% top_n(2, avg_logFC)

# Top markers plots
# Cluster 0
FeaturePlot(object = New, features.plot = c("Igf1r", "Spock2",'Adgrf5','Pdgfb','Slc39a10','Slc7a1'), cols.use = c("grey", "blue"),reduction.use = "tsne")
# Cluster 1
FeaturePlot(object = New, features.plot = c("Junb", "Jun",'Btg2','Fos','Atf3','Ier2'), cols.use = c("grey", "blue"),reduction.use = "tsne")
# Cluster 2
FeaturePlot(object = New, features.plot = c("Itm2a", "Nedd8",'Anp32a','Cbx3','Car4','Apoe'), cols.use = c("grey", "blue"),reduction.use = "tsne")
# Cluster 3
FeaturePlot(object = New, features.plot = c("Gkn3", "Glul",'Stmn2','Vegfc','Alpl','Clu'), cols.use = c("grey", "blue"),reduction.use = "tsne")
# Cluster 4
FeaturePlot(object = New, features.plot = c("Cd14", "Lcn2",'Lrg1','Vcam1','Ptgs2','Ackr1'), cols.use = c("grey", "blue"),reduction.use = "tsne")
# Cluster 5
FeaturePlot(object = New, features.plot = c("Vtn", "Atp1a2",'Rgs4','Kcnj8','Myl9','Pdgfrb'), cols.use = c("grey", "blue"),reduction.use = "tsne")
# Cluster 6
FeaturePlot(object = New, features.plot = c("Plvap", "Nrp1",'Car8','Esm1','Plpp3','Igfbp3'), cols.use = c("grey", "blue"),reduction.use = "tsne")
# Cluster 7
FeaturePlot(object = New, features.plot = c("Cd52", "Cd74",'Cd79a','Cd79b','Plac8','Lyz2'), cols.use = c("grey", "blue"),reduction.use = "tsne")


# Feature Plots for known markers
FeaturePlot(object = New, features.plot = c("Pecam1",'Cldn5','Efnb2','Dll4','Ephb4','Mfsd2a'), cols.use = c("grey", "blue"),reduction.use = "tsne")
FeaturePlot(object = New, features.plot = c("Pecam1",'Cldn5',"Vcam1",'Icam1','Lrg1','Cd74'), cols.use = c("grey", "blue"),reduction.use = "tsne")
FeaturePlot(object = New, features.plot = c("Pecam1",'Cldn5','Pvlap','Vwf','Ptprc','Cd14'), cols.use = c("grey", "blue"),reduction.use = "tsne")

# Violin Plots for cluster markers
VlnPlot(object = New, features.plot = c("Igf1r", "Spock2",'Adgrf5','Pdgfb','Slc39a10','Slc7a1'))
VlnPlot(object = New, features.plot = c("Junb", "Jun",'Btg2','Fos','Atf3','Ier2'))
VlnPlot(object = New, features.plot = c("Itm2a", "Nedd8",'Anp32a','Cbx3','Car4','Apoe'))
VlnPlot(object = New, features.plot = c("Gkn3", "Glul",'Stmn2','Vegfc','Alpl','Clu'))
VlnPlot(object = New, features.plot = c("Cd14", "Lcn2",'Lrg1','Vcam1','Ptgs2','Ackr1'))
VlnPlot(object = New, features.plot = c("Vtn", "Atp1a2",'Rgs4','Kcnj8','Myl9','Pdgfrb'))
VlnPlot(object = New, features.plot = c("Plvap", "Plpp1",'Car8','Esm1','Plpp3','Igfbp3'))
VlnPlot(object = New, features.plot = c("Cd52", "Cd74",'Cd79a','Cd79b','Plac8','Lyz2'))


# Plot heatmap with the top 10 genes for each cluster
top10 <- New.markers %>% group_by(cluster) %>% top_n(10, avg_logFC)
# setting slim.col.label to TRUE will print just the cluster IDS instead of
# every cell name
DoHeatmap(object = New, genes.use = top10$gene, slim.col.label = TRUE, remove.key = TRUE)

#FeaturePlot(object = New, features.plot = c("Junb", "Jun",'Jund','Fos','Cdkn1a','Ier2'), cols.use = c("grey", "blue"),reduction.use = "tsne")
#FeaturePlot(object = New, features.plot = c("Nfib",'Igf1r','Mcf2l','Kdr','Cxcl12','Hmcn1'), cols.use = c("grey", "blue"),reduction.use = "tsne")
#FeaturePlot(object = old, features.plot = c('Stmn2','Gkn3','Tm4sf1','Lypd1','Mgp','Clu'), cols.use = c("grey", "blue"),reduction.use = "tsne")
#FeaturePlot(object = New, features.plot = c('Fth1','Lcn2','Lrg1','Vcam1','Icam1','Ptgs2'), cols.use = c("grey", "blue"),reduction.use = "tsne")
#FeaturePlot(object = old, features.plot = c('Myl9','Atp1a2','Vtn','Cald1','Acta2','Rgs4'), cols.use = c("grey", "blue"),reduction.use = "tsne")
#FeaturePlot(object = old, features.plot = c("Junb", "Jun",'Jund','Fos','Adm','Btg2'), cols.use = c("grey", "blue"),reduction.use = "tsne")






















VlnPlot(object = New, features.plot = c("Junb", "Jun",'Jund','Fos','Ubb','Rpl41'))
VlnPlot(object = New, features.plot = c("Junb",'Rpl41','Igf1r','Ubb','Fos','Slco1c1'))
VlnPlot(object = New, features.plot = c('Lrg1','Ackr1','Ch25h','Lcn2','Cfh','Vcam1'))
VlnPlot(object = New, features.plot = c('Esm1','Ces2e','Fam167b','Car8','Col15a1','Plvap'))
top10 <- New.markers %>% group_by(cluster) %>% top_n(10, avg_logFC)
# setting slim.col.label to TRUE will print just the cluster IDS instead of
# every cell name
DoHeatmap(object = New, genes.use = top10$gene, slim.col.label = TRUE, remove.key = TRUE)

# Feature Plots
FeaturePlot(object = New, features.plot = c("Pecam1",'Cldn5','Efnb2','Dll4','Ephb4','Mfsd2a'), cols.use = c("grey", "blue"),reduction.use = "tsne")
FeaturePlot(object = New, features.plot = c("Pecam1",'Cldn5',"Vcam1",'Icam1','Lrg1','Cd74'), cols.use = c("grey", "blue"),reduction.use = "tsne")
FeaturePlot(object = New, features.plot = c("Pecam1",'Cldn5','Pvlap','Vwf','Ptprc','Cd14'), cols.use = c("grey", "blue"),reduction.use = "tsne")
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
